### PR TITLE
Fully implement V1 profiles#index

### PR DESCRIPTION
**Pull request by akofink**
Separated from https://github.com/RedHatInsights/compliance-backend/pull/410 to ease reviewing.

- Adds relationships to API response
- Allows side-loaded includes of all related resources
- Allows searching profiles by `canonical`

Signed-off-by: Andrew Kofink <akofink@redhat.com>

### CODE REVIEW

File: app/models/concerns/profile_searching.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    scope :has_test_results, lambda { |has_test_results = true|
+      test_results = TestResult.select(:profile_id).distinct
+      has_test_results && where(id: test_results) || where.not(id: test_results)
+    }
+  end
+
+  class_methods do
+    def canonical?(_filter, _operator, value)
+      profiles = Profile.canonical(ActiveModel::Type::Boolean.new.cast(value))
+      { conditions: profiles.arel.where_sql.gsub(/^where /i, '') }
```
Comment (akofink) [inline app/models/concerns/profile_searching.rb:33]: This is an attempt at making `scoped_search` tied to our `scope` for this model and search parameter. It seems silly to have the logic in two different places, but this solution is also kind of a hack.

File: app/models/concerns/profile_searching.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    scoped_search on: %i[id name ref_id account_id compliance_threshold
+                         external parent_profile_id]
+    scoped_search relation: :hosts, on: :id, rename: :system_ids
+    scoped_search relation: :hosts, on: :name, rename: :system_names
+    scoped_search on: :has_test_results, ext_method: 'test_results?',
+                  only_explicit: true, operators: ['=']
+    scoped_search on: :canonical, ext_method: 'canonical?', only_explicit: true,
+                  operators: ['=']
+
+    scope :canonical, lambda { |canonical = true|
```
Comment (akofink) [inline app/models/concerns/profile_searching.rb:17]: Notice the added param here: `Profile.canonical` or `Profile.canonical(true)` are equivalent. `Profile.canonical(false)` is the negative case. The same is true of `Profile.external` and `Profile.has_test_results`

File: app/serializers/parent_profile_serializer.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+# JSON API serialization for an OpenSCAP Profile
+class ParentProfileSerializer < ProfileSerializer
```
Comment (akofink) [inline app/serializers/parent_profile_serializer.rb:4]: This is required as a quirk of the way FastJsonapi works with `belongs_to :parent_profile`

File: spec/swagger_helper.rb [hunk 4]
```diff
@@ -117,11 +117,64 @@
           properties: {
```
Comment (akofink) [inline spec/swagger_helper.rb:117]: This ought to always match the parameters returned by the associated serializer. I wonder if we could add an automated test that ensures they match for all API resources.

File: app/controllers/concerns/rendering.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  extend ActiveSupport::Concern
+
+  included do
+    def render_json(model, **args)
+      opts = ['index'].include?(action_name) ? metadata : {}
+      opts.merge!(include: params[:include].split(',')) if params[:include]
+      render({ json: serializer.new(model, opts) }.merge(args))
+    end
+
+    def render_error(model, **args)
```
Comment (dLobatog) [inline app/controllers/concerns/rendering.rb:14]: Is this used anywhere?

Comment (akofink) [reply app/controllers/concerns/rendering.rb:14]: Good catch - this one was from create/update, but we'll be using it soon :)

**Pull request merged**